### PR TITLE
Implement ORDER BY optimization for Storage Merge, Buffer and MV

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
Implement `ORDER BY` optimization for `Merge`, `Buffer` and `Materilized View` storages with underlying MergeTree tables. 

Detailed description (optional):
Also refactor code near analyzing order description to make it more reusable for other expressions.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Also need to check for Buffer, View, MaterializedView.

**Comment by CurtizJ** [comment]
For storage View it works if it's saved query is like `SELECT ... FROM merge_tree_table ORDER BY pk`. But it's hard to support for queries like `SELECT ... FROM view ORDER BY pk` if saved query doesn't have `ORDER BY` clause, because View doesn't store storage of source table unlike to Buffer, MaterializedView and Merge storages.

File: dbms/src/Storages/StorageMerge.cpp [hunk 1]
```diff
@@ -217,15 +217,14 @@ BlockInputStreams StorageMerge::read(
             auto current_info = query_info.order_by_optimizer->analyze(it->first);
             if (it == selected_tables.begin())
                 input_sorting_info = current_info;
-            else if (!current_info || *current_info != *input_sorting_info)
+            else if (!current_info || (input_sorting_info && *current_info != *input_sorting_info))
```
Comment (CurtizJ) [inline dbms/src/Storages/StorageMerge.cpp]: It was PVS-studio false-positive.

File: dbms/src/Storages/ReadInOrderOptimizer.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    if (elements_actions.size() != required_sort_description.size())
+        throw Exception("Sizes of sort description and actions are mismatched", ErrorCodes::LOGICAL_ERROR);
+
+    /// Do not analyze ARRAY JOIN result columns.
+    /// TODO: forbid more columns for analyzing.
+    for (const auto & elem : syntax_result->array_join_result_to_source)
+        forbidden_columns.insert(elem.first);
+}
+
+InputSortingInfoPtr ReadInOrderOptimizer::analyze(const StoragePtr & storage) const
```
Comment (akuzm) [inline dbms/src/Storages/ReadInOrderOptimizer.cpp]: What does this function do, does it apply a projection (in relational sense) to a merge tree sorting key? E.g. you have some expressions derived from primary columns, you have the input sorting key in terms of primary columns, and you try to pull this key through the expressions and see what it becomes? Maybe we can think of a better name for the function based on this.

File: dbms/src/Storages/ReadInOrderOptimizer.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    const SortDescription & required_sort_description_,
+    const SyntaxAnalyzerResultPtr & syntax_result)
+    : elements_actions(elements_actions_)
+    , required_sort_description(required_sort_description_)
+{
+    if (elements_actions.size() != required_sort_description.size())
+        throw Exception("Sizes of sort description and actions are mismatched", ErrorCodes::LOGICAL_ERROR);
+
+    /// Do not analyze ARRAY JOIN result columns.
+    /// TODO: forbid more columns for analyzing.
```
Comment (akuzm) [inline dbms/src/Storages/ReadInOrderOptimizer.cpp]: Why do we need to specifically forbid some columns? We won't have them in merge tree sorting key anyway. This needs a comment.

Comment (CurtizJ) [reply dbms/src/Storages/ReadInOrderOptimizer.cpp]: Aliases to joined columns comes to order by as is and we can wrongly match them with order key of MergeTree.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
We can also add a perf test.

**Pull request merged**